src/build/cargo.rs

```diff
@@ -409,49 +409,59 @@ impl Executor for RlsExecutor {
-                    // as others may emit required metadata for dependent crate types
-                    if a.starts_with("--emit") && is_final_crate_type && !self.workspace_mode {
-                        "--emit=dep-info"
+            if config.all_targets {
+                if config.cfg_test {
+                    args.push("--cfg".to_owned());
+                    args.push("test".to_owned());
+                }
+            } else {
+                let crate_type = parse_arg(cargo_args, "--crate-type");
+
+                // Because we only try to emulate `cargo test` using `cargo check`, so for now
+                // assume crate_type arg (i.e. in `cargo test` it isn't specified for --test targets)
+                // and build test harness only for final crate type
+                let crate_type = crate_type.expect("no crate-type in rustc command line");
+                let build_lib = *config.build_lib.as_ref();
+                let is_final_crate_type = crate_type == "bin" || (crate_type == "lib" && build_lib);
+
+                if config.cfg_test {
+                    // FIXME(#351) allow passing --test to lib crate-type when building a dependency
+                    if is_final_crate_type {
+                        args.push("--test".to_owned());
                     } else {
-                        a
+                        args.push("--cfg".to_owned());
+                        args.push("test".to_owned());
                     }
-                })
-                .collect::<Vec<_>>();
-            cmd.args_replace(&modified);
+                }
+
+                // We can't omit compilation here, because Cargo is going to expect to get
+                // dep-info for this crate, so we shell out to rustc to get that.
+                // This is not really ideal, because we are going to
+                // compute this info anyway when we run rustc ourselves, but we don't do
+                // that before we return to Cargo.
+                // FIXME Don't do this. Start our build here rather than on another thread
+                // so the dep-info is ready by the time we return from this callback.
+                // NB: In `workspace_mode` regular compilation is performed here (and we don't
+                // only calculate dep-info) so it should fix the problem mentioned above.
+                let modified = args.iter()
+                    .map(|a| {
+                        // Emitting only dep-info is possible only for final crate type, as
+                        // as others may emit required metadata for dependent crate types
+                        if a.starts_with("--emit") && is_final_crate_type && !self.workspace_mode {
+                            "--emit=dep-info"
+                        } else {
+                            a
+                        }
+                    })
+                    .collect::<Vec<_>>();
+                cmd.args_replace(&modified);
+            };
```

Comment:
Hi @estin, thanks for the changes. It is clearer now what you are doing, but I'm afraid I still don't understand why. Could you explain why you have made these changes - what is it about --all-targets which means we need to do these things differently?

Reply:
Oh... I can't explain this logic. And I made new iteration for this PR and simplify introduction --all-targets flag again... I hope now it woulde be correct